### ISSUE TITLE
client.html paths are broken for npm@3

### ISSUE THREAD

**Issue by sholladay** [theintern/intern#508]
**Edit:** My assumptions were slightly off. This turned out to be an out-of-date symlink due to renaming the project directory. See my next comment.

A couple of weeks ago, npm@3 was released. It's greatness comes with some mayhem for pathing expectations, since it now flattens the dependency tree on disk, in most cases.

As a result, Intern's out-of-the-box client.html experience is completely broken due to 404s, in cases where the same project setup succeeds when installed with npm@2.

Assuming this URL points to a legit project ...
```
http://localhost/my-project/node_modules/intern/client.html?config=tests/intern
```
.. you will get a 404 for ...
```
http://localhost/my-project/node_modules/intern/node_modules/dojo/loader.js
```
... because that file now lives at ...
```
http://localhost/my-project/node_modules/dojo/loader.js
```
... completely outside of Intern's directory.

**Comment by csnover** [comment]
The purpose of fixdeps.js (install script) is to ensure that the dependencies are always available there. Maybe your Web server just is not configured to follow symlinks?

**Comment by sholladay** [comment]
Well, no, but that led me to the answer.

I had done the equivalent of:
```sh
mv my-project my-fav-project
```
So that symlink pointed to a disk location that no longer existed.

I'm not sure what I can really expect Intern to do about this. It's a bit premature to drop support for npm@2, but maybe the paths could be optimized for npm@3, with symlinks for where npm@2 expects them to be?

**Comment by csnover** [comment]
Options I can think of are:

* Use `path.relative`, which would make it OK for you to rename the project directory, but this won’t work for Windows (the junction points are absolute)
* Just make a copy, but that’s unpleasant and wasteful and could be a bummer if anyone runs dedupe (though to be honest I have no idea what happens now if someone tries to dedupe)
* Make a copy using hard links, which won’t waste space but is likely to confuse and confound and requires a lot of extra complexity (like recreating the entire directory structure)

**Comment by kitsonk** [comment]
I just tried npm@3, after I did a `rm -rf node_modules` and then an `npm install` and there were no issues in running the `client.html`.  Obviously the important thing is the `fixdeps.js` running when the package is installed.

**Comment by odoe** [comment]
I've run into issues when I include intern in a project that also does a Dojo build. In this case, I have a `node_modules/dojo` dep for my app using dojo 1.10. Intern has the same dep, but it's dojo2, and it looks like dojo 1.10 takes priority in that folder and breaks Intern. I've since reverted back to npm 2.x to avoid this issue on my machine.

You can see this happen with this repo https://github.com/odoe/esrijs-generator-demo

**Comment by sholladay** [comment]
To avoid muddying the topic or confusing people, I'm closing this. My original problem was not related to using npm@3, but due to re-naming my project directory, which I just happened to do right after installing npm@3 and before re-running the browser client. So there were two variables at play and it didn't occur to me that re-naming the directory would have caused this, given that I was using the correct up-to-date URL for the browser client. I can confirm that a clean install of Intern on npm@3 without doing that works for me in this scenario. I've opened a new issue over at #510 for the inability to re-name project directories, for more clarity. @odoe I suggest you do the same, in case it has a different cause.
